From b11e622f710b94792d2f678a90cbe034a08382c2 Mon Sep 17 00:00:00 2001
From: Saxena, Parth <parth.saxena@ti.com>
Date: Fri, 24 Jun 2011 13:19:45 +0530
Subject: [PATCH 16/20] ARM:omap:hsmmc: use clk_get_rate() instead of hard coding clock rate

Initially, the functional clock to the mmc controller was hard coded
as 96MHz in the driver. Thus, boards that did not have a functional
clock of 96MHz coming in the controller performed sub-optimally.
This patch removes the hard coded clock rate and obtaines the clock
rate specific to the board using clk_get_rate() function.

As a fallback, if the received clk rate is < hard coded value, then the
hard coded value is considered

Signed-off-by: Saxena, Parth <parth.saxena@ti.com>
Signed-off-by: Hebbar, Gururaja <gururaja.hebbar@ti.com>
---
 drivers/mmc/host/omap_hsmmc.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/mmc/host/omap_hsmmc.c b/drivers/mmc/host/omap_hsmmc.c
index 21e4a79..f26da0b 100644
--- a/drivers/mmc/host/omap_hsmmc.c
+++ b/drivers/mmc/host/omap_hsmmc.c
@@ -599,12 +599,17 @@ static void omap_hsmmc_disable_irq(struct omap_hsmmc_host *host)
 }
 
 /* Calculate divisor for the given clock frequency */
-static u16 calc_divisor(struct mmc_ios *ios)
+static u16 calc_divisor(struct mmc_ios *ios, struct omap_hsmmc_host *host)
 {
 	u16 dsor = 0;
+	u32 clk_rate;
 
 	if (ios->clock) {
-		dsor = DIV_ROUND_UP(OMAP_MMC_MASTER_CLOCK, ios->clock);
+		clk_rate = clk_get_rate(host->fclk);
+		if (clk_rate < OMAP_MMC_MASTER_CLOCK)
+			clk_rate = OMAP_MMC_MASTER_CLOCK;
+
+		dsor = DIV_ROUND_UP(clk_rate, ios->clock);
 		if (dsor > 250)
 			dsor = 250;
 	}
@@ -624,7 +629,7 @@ static void omap_hsmmc_set_clock(struct omap_hsmmc_host *host)
 
 	regval = OMAP_HSMMC_READ(host->base, SYSCTL);
 	regval = regval & ~(CLKD_MASK | DTO_MASK);
-	regval = regval | (calc_divisor(ios) << 6) | (DTO << 16);
+	regval = regval | (calc_divisor(ios, host) << 6) | (DTO << 16);
 	OMAP_HSMMC_WRITE(host->base, SYSCTL, regval);
 	OMAP_HSMMC_WRITE(host->base, SYSCTL,
 		OMAP_HSMMC_READ(host->base, SYSCTL) | ICE);
-- 
1.6.6.1

